### PR TITLE
fix: union types of union types in Records

### PR DESCRIPTION
**Pull request by domoritz**
Revised version of https://github.com/vega/ts-json-schema-generator/pull/1926 by [dpvball](https://github.com/dpvball)

<details>
  <summary>📦 Published PR as canary version: <code>2.1.1--canary.1933.1cab987.0</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install ts-json-schema-generator@2.1.1--canary.1933.1cab987.0
  # or 
  yarn add ts-json-schema-generator@2.1.1--canary.1933.1cab987.0
  ```
</details>

### CODE REVIEW

**Comment by dpvball** [comment]
@domoritz Much more eloquent solution than my PR. Didn't even notice the existence of the derefType utility. 

Few comments:
- I tested it with your code. It looks like it only fails when you have a duplicate reference like we have in the test example. Since MyType3 references MyType2 and MyType2 is already referenced in MyType10. So s2,s3 will show up multiple times in the flattenedType array. I believe you will need to include the duplicate removal code after the LiteralType filter in MappedTypedNodeParser. Not sure if you have a utility function that can more cleanly remove duplicates. But i'll just post what i had before there.
- I added the allowUnionType since running the tests result in this warning. If that's fine. no issues leaving that out. `strict mode: use allowUnionTypes to allow union type keyword at "#/properties/nullableValue" (strictTypes)`
- I think you missed bringing over the test in test/valid-data-type.test.ts

thanks again for all the help on this. I use union types in records extensively in my code and this will help greatly.

File: src/NodeParser/MappedTypeNodeParser.ts [hunk 1]
```diff
@@ -103,7 +103,7 @@ export class MappedTypeNodeParser implements SubNodeParser {
 
     protected getProperties(node: ts.MappedTypeNode, keyListType: UnionType, context: Context): ObjectProperty[] {
         return keyListType
-            .getTypes()
+            .getFlattenedTypes(derefType)
             .filter((type): type is LiteralType => type instanceof LiteralType)
             .map((type) => [type, this.mapKey(node, type, context)])
```
Comment (dpvball) [inline src/NodeParser/MappedTypeNodeParser.ts:104]:
```suggestion
            .reduce((acc: LiteralType[], curr: LiteralType) => {
                if (
                    acc.findIndex((val: LiteralType) => {
                        return val.getId() === curr.getId();
                    }) < 0
                ) {
                    acc.push(curr);
                }
                return acc;
            }, [])
            .map((type) => [type, this.mapKey(node, type, context)])
```

Comment (domoritz) [reply src/NodeParser/MappedTypeNodeParser.ts:104]: We can use `uniqueTypeArray` here instead I think.

**Comment by dpvball** [comment]
thanks for looking into this so quickly @domoritz

**Pull request merged**